src/core/service.c

```diff
@@ -252,6 +252,26 @@ static void service_start_watchdog(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to install watchdog timer: %m");
 }
 
+static void service_extend_timeout(Service *s) {
+        assert(s);
+
+        if (s->timer_event_source) {
+                uint64_t current, extended;
+                int r;
+
+                r = sd_event_source_get_time(s->timer_event_source, &current);
+                if (r < 0)
+                        log_unit_warning_errno(UNIT(s), r, "Failed to retrieve timeout timer: %m");
```

Comment:
hmm, i'd just fail this function if sd_event_source_get_time returns < 0. Use log_unit_error_errno() and return r. This is not going to happen anyway, and if it does, then there's something seriously wrong and it's probably better to not restart the timeouts with current == 0.